### DISCUSSION TITLE
Explicit implementation of interface 'init' property accessor

### DISCUSSION THREAD

**Discussion by biqas** [dotnet/csharplang#4519] [Q&A] [answered]
Hi,

I hit the following limitation for initializing an explicitly implemented interface init property.

```csharp
public class UserRoleEntity : IUserRole
{
    public UserRoleEntity(IUser user, IRole role)
    {
        ((IUserRole)this).User = user; // this here does not work, described above *.
        this.Role = role; // this here does not work because IRole not RoleEntity.
    }

    public UserEntity User { get; init; }
    public RoleEntity Role { get; init; }

    IUser IUserRole.User { get; init; }
    IRole IUserRole.Role { get; init; }
}

public interface IUserRole
{
    IUser User { get; init; }
    IRole Role { get; init; }
}
```
Is there a possibility to initialize the property?

\* Init-only property or indexer can only be assigned in an object initializer, or on 'this' or 'base' in an instance constructor or an 'init' accessor.

**Comment by jnm2** [comment]
Side question, is it really preferable to allow the explicitly implemented properties to be set to something different than the public properties? I would have anticipated seeing this:

```cs
    public UserEntity User { get; init; }
    public RoleEntity Role { get; init; }

    IUser IUserRole.User => User;
    IRole IUserRole.Role => Role;
 ```

**Reply by biqas** [reply]
@jnm2 
Because I'm in control of this I have already changed to the same one you are describing, but what would I do if not in control??

**Reply by jnm2** [reply]
@biqas If you find yourself needing to set explicitly implemented properties that were created with `init` accessors, you'd need to inform whoever created the class that C# doesn't allow such properties to be set.

**Comment by 333fred** [comment, marked as answer]
We explicitly looked at this, and rejected the feature: https://github.com/dotnet/csharplang/blob/master/meetings/2021/LDM-2021-01-27.md#init-only-access-on-conversion-on-this.
